yacl/crypto/primitives/psu/krtw19_psu.cc

```diff
@@ -0,0 +1,241 @@
+// Copyright 2024 zhangwfjh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "yacl/crypto/primitives/psu/krtw19_psu.h"
+
+#include <algorithm>
+#include <array>
+#include <iterator>
+#include <random>
+
+#include "yacl/crypto/base/hash/hash_utils.h"
+#include "yacl/crypto/primitives/ot/base_ot.h"
+#include "yacl/crypto/primitives/ot/iknp_ote.h"
+#include "yacl/crypto/primitives/ot/kkrt_ote.h"
+#include "yacl/crypto/utils/rand.h"
+#include "yacl/math/gadget.h"
+#include "yacl/utils/serialize.h"
+
+namespace yacl::crypto::psu {
+
+namespace yc = yacl::crypto;
+
+namespace {
+
+// reference: https://eprint.iacr.org/2019/1234.pdf (Figure 2)
+constexpr float ZETA{0.06f};
+constexpr size_t BIN_SIZE{64ul};  // m+1
+constexpr uint128_t BOT{};
+constexpr size_t NUM_BIN_PER_BATCH{16ul};
+constexpr size_t BATCH_SIZE{NUM_BIN_PER_BATCH * BIN_SIZE};
+
+constexpr size_t NUM_BASE_OT{128ul};
+constexpr size_t NUM_INKP_OT{512ul};
+
+static std::random_device rd;
+static std::mt19937 gen(rd());
+
+struct HashU128 {
+  size_t operator()(const uint128_t& x) const {
+    return yacl::math::UniversalHash<uint128_t>(1, {x});
```

Comment:
P0: Please use the collision-resistent hash function (Blake3, Sha2, ...), which are defined in `yacl/crypto/base/hash/hash_utils.h`.

FYI, the universal hash is considered weaker (but faster) than the CRHash, see: https://crypto.stackexchange.com/a/88247/61581 for more discussions of the security model